packages/master/master.js

```diff
@@ -690,12 +691,22 @@ class ControlConnection extends BaseConnection {
 	async listInstancesRequestHandler(message) {
 		let list = [];
 		for (let instance of db.instances.values()) {
-			list.push({
+			let push_info = {
 				id: instance.config.get("instance.id"),
 				name: instance.config.get("instance.name"),
 				assigned_slave: instance.config.get("instance.assigned_slave"),
+				assigned_slave_name: null,
+				game_port: instance.config.get("factorio.game_port"),
+				public_address: null,
 				status: instance.status,
-			});
+			};
+			// get assigned slave info
+			if (instance.config.get("instance.assigned_slave") !== null) {
+				let assigned_slave_info = db.slaves.get(instance.config.get("instance.assigned_slave"));
+				push_info.assigned_slave_name = assigned_slave_info.name;
+				push_info.public_address = assigned_slave_info.public_address;
```

Comment:
`assigned_slave_info` may be `undefined` here if for example `database/slaves.json` is deleted, rendering the cluster unusable.

Reply:
i think `assigned_slave_info` defalut is `undefined` is not problem, if player create new instance and don't assgin any slave, the `assigned_slave_info` is`undefined`, in WebUI, `public_address` and `assigned_slave_name` is display `undefined`, but instance must be assigned a slave, after the assign to any slave, `assigned_slave_info` can get slave data

Reply:
If `assigned_slave_info` is `undefined` then `assigned_slave_info.name` throws `TypeError: Cannot read property 'name' of undefined`.

Reply:
`if (instance.config.get("instance.assigned_slave") !== null) {` determine whether the instance is assigned to a slave, so  don't worry about this problem

Reply:
There is no guarantee that the id returned by `instance.config.get("instance.assigned_slave")` is valid, and it's easy enough to get into a situation where it's invalid by deleting database/slave.json.